File: tests/test_v1172a_x_widget.php
<?php

declare(strict_types=1);

$cacheRoot = sys_get_temp_dir() . '/rss-reader-v1172a-x-' . bin2hex(random_bytes(6));
define('APP_X_BEARER_TOKEN', 'v1172a-test-token-not-secret');
define('APP_X_CACHE_TTL_SECONDS', 300);
define('APP_X_STALE_MAX_AGE_SECONDS', 3600);
define('APP_X_TIMEOUT_MS', 5000);
define('APP_X_CACHE_DIR', $cacheRoot);
define('APP_ENV', 'testing');
define('APP_DEBUG', false);
define('APP_LOG_ENABLED', false);

require dirname(__DIR__) . '/app/bootstrap.php';

$pass = 0;
$fail = 0;

function v1172a_check(bool $condition, string $message): void
{
    global $pass, $fail;
    if ($condition) {
        $pass++;
        echo "PASS: {$message}\n";
        return;
    }
    $fail++;
    echo "FAIL: {$message}\n";
}

function v1172a_rrmdir(string $dir): void
{
    if (!is_dir($dir)) {
        return;
    }
    foreach (scandir($dir) ?: [] as $name) {
        if ($name === '.' || $name === '..') {
            continue;
        }
        $path = $dir . '/' . $name;
        is_dir($path) ? v1172a_rrmdir($path) : @unlink($path);
    }
    @rmdir($dir);
}

try {
    v1172a_check(in_array('x_timeline', dashboard_widget_types(), true), 'dashboard widget type includes x_timeline');
    v1172a_check(in_array('x_timeline', information_widget_types(), true), 'shared information persistence accepts x_timeline');

    v1172a_check(x_widget_validate_username('@XDevelopers') === 'XDevelopers', 'username accepts leading @ and normalizes it');
    v1172a_check(x_widget_validate_username('abc_123') === 'abc_123', 'username accepts X handle characters');
    v1172a_check(x_widget_validate_username('bad-handle') === null, 'username rejects unsupported characters');
    v1172a_check(x_widget_validate_username(str_repeat('a', 16)) === null, 'username rejects more than 15 characters');

    $config = x_widget_config_from_input([
        'x_title' => 'X News',
        'x_username' => '@XDevelopers',
        'x_display_count' => '3',
        'x_show_replies' => '0',
        'x_show_reposts' => '0',
    ]);
    v1172a_check(is_array($config), 'valid X widget config is normalized');
    v1172a_check(($config['username'] ?? null) === 'XDevelopers', 'normalized config stores username without @');
    v1172a_check(($config['display_count'] ?? null) === 3, 'display count 3 is accepted');
    v1172a_check(($config['show_replies'] ?? true) === false && ($config['show_reposts'] ?? true) === false, 'reply/repost filters are normalized');
    v1172a_check(x_widget_validate_display_count('4') === null, 'unsupported display count is rejected');

    foreach ([401 => 'x_auth_failed', 403 => 'x_access_forbidden', 404 => 'x_not_found', 429 => 'x_rate_or_usage_limited'] as $status => $reason) {
        $actualReason = '';
        try {
            x_api_request_json('https://api.x.com/2/test', static fn(string $url, string $token): array => [
                'ok' => true, 'status' => $status, 'body' => '{\"title\":\"test\"}',
            ]);
        } catch (XApiRequestException $exception) {
            $actualReason = $exception->reasonCode();
        }
        v1172a_check($actualReason === $reason, 'HTTP ' . $status . ' maps to ' . $reason);
    }

    $calls = [];
    $fetcher = static function (string $url, string $token) use (&$calls): array {
        $calls[] = ['url' => $url, 'token' => $token];
        if (str_contains($url, '/2/users/by/username/')) {
            return [
                'ok' => true,
                'status' => 200,
                'body' => json_encode(['data' => [
                    'id' => '2244994945',
                    'name' => 'X Developers',
                    'username' => 'XDevelopers',
                    'protected' => false,
                ]], JSON_UNESCAPED_SLASHES),
            ];
        }
        if (str_contains($url, '/2/users/2244994945/tweets')) {
            $posts = [];
            for ($i = 1; $i <= 5; $i++) {
                $posts[] = [
                    'id' => (string) (1000000000000000000 + $i),
                    'text' => 'Post ' . $i,
                    'created_at' => sprintf('2026-08-19T0%d:00:00Z', $i),
                ];
            }
            return [
                'ok' => true,
                'status' => 200,
                'body' => json_encode(['data' => $posts], JSON_UNESCAPED_SLASHES),
            ];
        }
        return ['ok' => false, 'status' => 500, 'body' => ''];
    };

    /** @var array<string,mixed> $config */
    $timeline = x_widget_fetch_timeline($config, true, $fetcher);
    v1172a_check(count($timeline['posts'] ?? []) === 3, 'display count 3 returns only three posts');
    v1172a_check(($timeline['account']['username'] ?? null) === 'XDevelopers', 'timeline returns normalized public account data');
    v1172a_check(($timeline['stale'] ?? null) === false, 'successful API result is not stale');

    $timelineCalls = array_values(array_filter($calls, static fn(array $call): bool => str_contains($call['url'], '/tweets')));
    $timelineUrl = (string) ($timelineCalls[0]['url'] ?? '');
    $query = [];
    parse_str((string) parse_url($timelineUrl, PHP_URL_QUERY), $query);
    v1172a_check(($query['max_results'] ?? null) === '5', 'X API minimum max_results is five when widget displays three');
    v1172a_check(($query['post_fields'] ?? $query['post.fields'] ?? null) === 'created_at', 'timeline requests created_at via post.fields');
    v1172a_check(($query['exclude'] ?? null) === 'replies,retweets', 'timeline excludes replies and retweets by default');
    v1172a_check(!str_contains(json_encode($timeline, JSON_UNESCAPED_SLASHES) ?: '', APP_X_BEARER_TOKEN), 'Bearer Token is not included in normalized timeline output');
    v1172a_check(count($calls) === 2 && $calls[0]['token'] === APP_X_BEARER_TOKEN && $calls[1]['token'] === APP_X_BEARER_TOKEN, 'Bearer Token is used only by server-side fetch callback');

    $cachedCalls = 0;
    $cachedFetcher = static function () use (&$cachedCalls): array {
        $cachedCalls++;
        return ['ok' => false, 'status' => 500, 'body' => ''];
    };
    $cached = x_widget_fetch_timeline($config, false, $cachedFetcher);
    v1172a_check($cachedCalls === 0, 'fresh timeline cache avoids another X API request');
    v1172a_check(($cached['posts'][0]['text'] ?? null) === 'Post 1', 'fresh timeline cache preserves posts');

    $failingFetcher = static fn(string $url, string $token): array => [
        'ok' => false,
        'status' => 0,
        'body' => '',
        'error_code' => 'timeout',
        'error_message' => 'test timeout',
    ];
    $fallback = x_widget_fetch_timeline($config, true, $failingFetcher);
    v1172a_check(($fallback['stale'] ?? null) === true, 'API failure falls back to cached timeline and marks it stale');
    v1172a_check(($fallback['posts'][2]['text'] ?? null) === 'Post 3', 'stale fallback keeps the last successful posts');

    $forbiddenReason = '';
    $forbiddenFetcher = static fn(string $url, string $token): array => [
        'ok' => true,
        'status' => 403,
        'body' => '{\"title\":\"Forbidden\"}',
    ];
    try {
        x_widget_fetch_timeline($config, true, $forbiddenFetcher);
    } catch (XApiRequestException $exception) {
        $forbiddenReason = $exception->reasonCode();
    }
    v1172a_check($forbiddenReason === 'x_access_forbidden', 'authorization/access failure does not expose stale cached posts');

    $requestCaptured = null;
    $resolver = static fn(string $host): array => $host === 'api.x.com' ? ['104.244.42.193'] : [];
    $transport = static function (array $request) use (&$requestCaptured): array {
        $requestCaptured = $request;
        return [
            'ok' => true,
            'status' => 200,
            'body' => '{"data":{}}',
            'error_code' => '',
            'error_message' => '',
        ];
    };
    $safe = x_api_safe_fetch('https://api.x.com/2/users/by/username/XDevelopers', APP_X_BEARER_TOKEN, $resolver, $transport);
    v1172a_check(($safe['ok'] ?? false) === true, 'safe X fetch accepts only the fixed HTTPS API host');
    v1172a_check(is_array($requestCaptured) && ($requestCaptured['authorization_bearer'] ?? null) === APP_X_BEARER_TOKEN, 'safe transport receives explicit Bearer field');
    v1172a_check(is_array($requestCaptured) && ($requestCaptured['host'] ?? null) === 'api.x.com' && ($requestCaptured['ip'] ?? null) === '104.244.42.193', 'safe transport keeps DNS-pinned X target');
    $blocked = x_api_safe_fetch('https://example.com/2/users/1/tweets', APP_X_BEARER_TOKEN, $resolver, $transport);
    v1172a_check(($blocked['ok'] ?? true) === false && ($blocked['error_code'] ?? null) === 'invalid_target', 'safe X fetch rejects non-X hosts');

    $protectedFetcher = static function (string $url, string $token): array {
        return [
            'ok' => true,
            'status' => 200,
            'body' => json_encode(['data' => [
                'id' => '123456789',
                'name' => 'Private Account',
                'username' => 'PrivateTest',
                'protected' => true,
            ]], JSON_UNESCAPED_SLASHES),
        ];
    };
    $protectedConfig = $config;
    $protectedConfig['username'] = 'PrivateTest';
    $protectedReason = '';
    try {
        x_widget_fetch_timeline($protectedConfig, true, $protectedFetcher);
    } catch (XApiRequestException $exception) {
        $protectedReason = $exception->reasonCode();
    }
    v1172a_check($protectedReason === 'x_protected_account', 'protected accounts return a specific app-only access error');
} finally {
    v1172a_rrmdir($cacheRoot);
}

echo "SUMMARY PASS={$pass} FAIL={$fail}\n";
exit($fail === 0 ? 0 : 1);

File: tests/test_v1172a_x_widget.py
from pathlib import Path
import re
import sys

ROOT = Path(__file__).resolve().parents[1]
PASS = 0
FAIL = 0


def check(condition: bool, message: str) -> None:
    global PASS, FAIL
    if condition:
        PASS += 1
        print(f"PASS: {message}")
    else:
        FAIL += 1
        print(f"FAIL: {message}")


def text(path: str) -> str:
    return (ROOT / path).read_text(encoding="utf-8")

x_php = text("app/x_widget.php")
api_php = text("app/api.php")
http_php = text("app/http_fetch.php")
dash_php = text("app/dashboard_widget.php")
info_php = text("app/information_widget.php")
conf_php = text("app/common/common_conf.php")
version_php = text("app/version.php")
calendar_js = text("public/js/calendar.js")
x_js = text("public/js/x-widget.js")
x_css = text("public/css/x-widget.css")
local_example = text("config/local.php.example")
env_example = text("config/.env.example")
schema = text("database/schema.sql")

check("'x_timeline'" in dash_php, "dashboard widget type includes x_timeline")
check("'x_timeline'" in info_php, "shared information widget persistence accepts x_timeline")
check("widget.x.create" in api_php and "widget.x.update" in api_php and "widget.x.delete" in api_php, "X widget CRUD API actions are registered")
check("x.timeline.fetch" in api_php, "X timeline fetch API action is registered")
check("X_API_HOST = 'api.x.com'" in x_php, "server client pins X API host")
check("https://' . X_API_HOST . '/2/users/by/username/" in x_php, "username lookup uses X API v2 endpoint")
check("'/2/users/' . rawurlencode($account['id']) . '/tweets?'" in x_php, "user posts endpoint is built from validated numeric user id")
check("'max_results' => $maxResults" in x_php and "'post.fields' => 'created_at'" in x_php, "timeline requests bounded post fields and max_results")
check("$exclude[] = 'replies'" in x_php and "$exclude[] = 'retweets'" in x_php, "reply and repost filters use X timeline exclude values")
check("APP_X_CACHE_TTL_SECONDS" in conf_php and "APP_X_STALE_MAX_AGE_SECONDS" in conf_php and "APP_X_TIMEOUT_MS" in conf_php, "X timeout/cache settings have bounded server defaults")
check("authorization_bearer" in http_php and "Authorization: Bearer " in http_php, "safe HTTP transport has a narrow explicit Bearer path")
check("request_headers" in http_php and "If-None-Match" in http_php, "existing arbitrary-header restriction remains in safe HTTP transport")
check("api.x.com" not in x_js and "Authorization: Bearer" not in x_js and "replace-with-your-x-api-bearer-token" not in x_js, "browser module contains no X API endpoint or credential value/header")
check("./api_v1.php" in x_js, "browser X data requests stay on the local API")
check("window.location.reload" not in x_js, "X widget CRUD and refresh do not force a whole-page reload")
check("function insertCardSorted" in x_js and ".insertBefore(this)" in x_js, "X widget insertion targets one location without rebuilding the full grid")
check(".replaceWith($card)" in x_js, "X settings update replaces only the target X card")
check("$('<div>').addClass('x-widget-post-text').text" in x_js, "post text is rendered with text() rather than HTML")
check("/^https:\\/\\/x\\.com\\//i.test(url)" in x_js, "post links are restricted to x.com before rendering")
check("x-widget.js?v=1.17.2-a" in calendar_js and "x-widget.css?v=1.17.2-a" in calendar_js, "staged X assets are loaded with the V1.17.2-A revision")
check("APP_ASSET_REVISION = '1.17.2-a'" in version_php and "APP_VERSION = '1.17.1'" in version_php, "staged asset revision changes without prematurely finalizing APP_VERSION")
check("replace-with-your-x-api-bearer-token" in local_example and "replace-with-your-x-api-bearer-token" in env_example, "configuration examples contain placeholders only")
check("APP_X_BEARER_TOKEN" in local_example and "APP_X_BEARER_TOKEN" in env_example, "both supported private config examples document the Bearer Token")
check("connect-src" not in text("public/.htaccess") if (ROOT / "public/.htaccess").exists() else True, "V1.17.2-A does not require browser CSP access to api.x.com")

match = re.search(r"`widget_type` VARCHAR\((\d+)\)", schema)
check(bool(match) and len("x_timeline") <= int(match.group(1)), "x_timeline fits the existing dashboard_widget.widget_type column")
check("CREATE TABLE" not in x_php and "ALTER TABLE" not in x_php, "X widget runtime adds no schema mutation")
check("APP_X_BEARER_TOKEN' => 'replace-with-your-x-api-bearer-token'" in local_example, "real Bearer Token is not embedded in the production example")
check(".x-widget-card" in x_css and "var(--bs-body-bg" in x_css and "var(--bs-body-color" in x_css, "X card styling follows Bootstrap theme variables")
check("x_protected_account" in x_php and "x_rate_or_usage_limited" in api_php, "public-app-only access and usage-limit errors are explicit")
check("function x_widget_exception_allows_stale" in x_php and "x_access_forbidden" not in x_php.split("function x_widget_exception_allows_stale", 1)[1].split("/** @param", 1)[0], "stale fallback is limited to transient failures, not authorization failures")

print(f"SUMMARY PASS={PASS} FAIL={FAIL}")
sys.exit(0 if FAIL == 0 else 1)
